consensus/hotstuff/cruisectl/block_time_controller_test.go

```diff
@@ -223,59 +216,89 @@ func (bs *BlockTimeControllerSuite) TestInit_EpochSetupPhase() {
+	require.NoError(bs.T(), err)
+
+	assert.Equal(bs.T(), extensionTargetTime, bs.ctl.curEpochTargetEndTime)
+	assert.Equal(bs.T(), extensionTargetDuration, bs.ctl.curEpochTargetDuration)
+	assert.Equal(bs.T(), extensionFinalView, bs.ctl.curEpochFinalView)
+
+	// duplicate events should be no-ops
+	for i := 0; i <= cap(bs.ctl.epochEvents); i++ {
+		bs.ctl.EpochExtended(bs.epochCounter, header, extension)
 	}
-	// state should be unchanged
-	now := time.Now().UTC()
-	assert.Equal(bs.T(), now.Add(bs.config.FallbackProposalDelay.Load()), bs.ctl.getProposalTiming().TargetPublicationTime(12, now, unittest.IdentifierFixture()))
+	require.Eventually(bs.T(), func() bool {
+		return len(bs.ctl.epochEvents) == 0
+	}, time.Second, time.Millisecond)
 
-	// additional OnBlockIncorporated events should be no-ops
-	for i := 0; i <= cap(bs.ctl.incorporatedBlocks); i++ {
-		header := unittest.BlockHeaderFixture(unittest.HeaderWithView(bs.initialView + 1 + uint64(i)))
-		header.ParentID = unittest.IdentifierFixture()
-		bs.ctl.OnBlockIncorporated(model.BlockFromFlow(header))
+	assert.Equal(bs.T(), extensionTargetTime, bs.ctl.curEpochTargetEndTime)
+	assert.Equal(bs.T(), extensionFinalView, bs.ctl.curEpochFinalView)
+}
+
+// TestOnEpochCommittedPhaseStarted ensures that the epoch info is updated when the next epoch is committed.
+func (bs *BlockTimeControllerSuite) TestOnEpochCommittedPhaseStarted() {
+	nextEpoch := mockprotocol.NewEpoch(bs.T())
+	nextEpoch.On("Counter").Return(bs.epochCounter+1, nil)
+	nextEpoch.On("FinalView").Return(bs.curEpochFinalView*2, nil)
+	nextEpoch.On("TargetDuration").Return(bs.EpochDurationSeconds(), nil)
+	nextEpoch.On("TargetEndTime").Return(bs.curEpochTargetEndTime+bs.EpochDurationSeconds(), nil)
+	bs.epochs.Add(nextEpoch)
+	bs.CreateAndStartController()
+	defer bs.StopController()
+	header := unittest.BlockHeaderFixture()
+	bs.ctl.EpochCommittedPhaseStarted(bs.epochCounter, header)
+	require.Eventually(bs.T(), func() bool {
+		return bs.ctl.nextEpochFinalView != nil
+	}, time.Second, time.Millisecond)
+
+	assert.Equal(bs.T(), bs.curEpochFinalView*2, *bs.ctl.nextEpochFinalView)
+	assert.Equal(bs.T(), bs.curEpochTargetEndTime+bs.EpochDurationSeconds(), *bs.ctl.nextEpochTargetEndTime)
+
+	// duplicate events should be no-ops
+	for i := 0; i <= cap(bs.ctl.epochEvents); i++ {
+		bs.ctl.EpochCommittedPhaseStarted(bs.epochCounter, header)
 	}
-	// wait for the channel to drain, since OnBlockIncorporated doesn't block on sending
+
 	require.Eventually(bs.T(), func() bool {
-		return len(bs.ctl.incorporatedBlocks) == 0
+		return len(bs.ctl.epochEvents) == 0
 	}, time.Second, time.Millisecond)
-	// state should be unchanged
-	now = time.Now().UTC()
-	assert.Equal(bs.T(), now.Add(bs.config.FallbackProposalDelay.Load()), bs.ctl.getProposalTiming().TargetPublicationTime(17, now, unittest.IdentifierFixture()))
+
+	assert.Equal(bs.T(), bs.curEpochFinalView*2, *bs.ctl.nextEpochFinalView)
```

Comment:
Could we also check that `bs.ctl.curEpochTargetDuration` has the expected value?